library/kotlin/src/io/envoyproxy/envoymobile/GRPCResponseHandler.kt

```diff
@@ -0,0 +1,177 @@
+    }
+
+    return this
+  }
+
+  /**
+   * Specify a callback for when trailers are received by the stream.
+   * If the closure is called, the stream is complete.
+   *
+   * @param closure: Closure which will receive the trailers.
+   * @return ResponseHandler, this ResponseHandler.
+   */
+  fun onTrailers(closure: (trailers: Map<String, List<String>>) -> Unit): GRPCResponseHandler {
+    underlyingHandler.onTrailers(closure)
+    return this
+  }
+
+  /**
+   * Specify a callback for when an internal Envoy exception occurs with the stream.
+   * If the closure is called, the stream is complete.
+   *
+   * @param closure: Closure which will be called when an error occurs.
+   * @return ResponseHandler, this ResponseHandler.
+   */
+  fun onError(closure: (error: EnvoyError) -> Unit): GRPCResponseHandler {
+    this.errorClosure = closure
+    underlyingHandler.onError(closure)
+    return this
+  }
+
+  /**
+   * Recursively processes a buffer of data, buffering it into messages based on state.
+   * When a message has been fully buffered, `onMessage` will be called with the message.
+   *
+   * @param bufferedStream The buffer of data from which to determine state and messages.
+   * @param processState The current process state of the buffering.
+   * @param onMessage Closure to call when a new message is available.
+   */
+  private fun processData(
+      bufferedStream: ByteArrayOutputStream,
+      processState: ProcessState,
+      onMessage: (byteBuffer: ByteBuffer) -> Unit): ProcessState {
+
+    var nextState = processState
+
+    when (processState) {
+      is ProcessState.CompressionFlag -> {
+        val byteArray = bufferedStream.toByteArray()
+        if (byteArray.isEmpty()) {
+          // We don't have enough information to extract the compression flag, so we'll just return
+          return ProcessState.CompressionFlag
+        }
+
+        val compressionFlag = byteArray[0]
+        // TODO: Support gRPC compression https://github.com/lyft/envoy-mobile/issues/501
+        if (compressionFlag.compareTo(0) != 0) {
+          errorClosure(
+              EnvoyError(
+                  EnvoyErrorCode.ENVOY_UNDEFINED_ERROR,
+                  "Unable to accept compression enabled messages for gRPC"))
+
+          // no op the current onData and clean up
+          underlyingHandler.onData { _, _ -> }
+          bufferedStream.reset()
+        }
+
+        nextState = ProcessState.MessageLength
+      }
+      is ProcessState.MessageLength -> {
+        if (bufferedStream.size() < GRPC_PREFIX_LENGTH) {
+          // We don't have enough information to extract the message length, so we'll just return
+          return ProcessState.MessageLength
+        }
+
+        val byteArray = bufferedStream.toByteArray()
+        val messageLength = ByteBuffer.wrap(byteArray.sliceArray(1..4)).int
+
+        if (messageLength == 0) {
+          onMessage(ByteBuffer.wrap(ByteArray(0)))
+          return ProcessState.CompressionFlag
```

Comment:
Because this is early returning instead of continuing, don't we need to drop the first 5 bytes of the `bufferedStream` here to unblock future responses?

Reply:
Hm, let me add a test

Reply:
It's also leaking message parsing responsibilities onto the message length parsing so I can move this on the `ProcessState.Message` statement